cluster/images/conformance/go-runner/run-conformance.go

```diff
@@ -0,0 +1,341 @@
+
+	gzw := gzip.NewWriter(outfile)
+	defer gzw.Close()
+
+	tw := tar.NewWriter(gzw)
+	defer tw.Close()
+
+	return filepath.Walk(dir, func(file string, fi os.FileInfo, err error) error {
+		// Return on any error.
+		if err != nil {
+			return err
+		}
+
+		// Only write regular files.
+		if !fi.Mode().IsRegular() || filepath.Join(dir, fi.Name()) == outpath {
+			return nil
+		}
+
+		// Create a new dir/file header.
+		header, err := tar.FileInfoHeader(fi, fi.Name())
+		if err != nil {
+			return errors.Wrap(err, "creating file info header for tarball")
+		}
+
+		// Update the name to correctly reflect the desired destination when untaring.
+		header.Name = strings.TrimPrefix(strings.Replace(file, dir, "", -1), string(filepath.Separator))
+		if err := tw.WriteHeader(header); err != nil {
+			return errors.Wrap(err, "writing header for tarball")
+		}
+
+		// Open files, copy into tarfile, and close.
+		f, err := os.Open(file)
+		if err != nil {
+			return errors.Wrap(err, "opening file for writing into tarball")
+		}
+
+		if _, err := io.Copy(tw, f); err != nil {
+			f.Close()
+			return errors.Wrap(err, "creating file contents into tarball")
+		}
+		f.Close()
+
+		return nil
+	})
+}
+
+// getCmd uses the given environment to form the ginkgo command to run tests. It will
+// set the stdout/stderr to the given writer.
+func getCmd(env Getenver, w io.Writer) *exec.Cmd {
+	ginkgoArgs := []string{}
+
+	// The logic of the parallel env var impacting the skip value necessitates it
+	// being placed before the rest of the flag resolution.
+	skip := env.Getenv(skipEnvKey)
+	switch env.Getenv(parallelEnvKey) {
+	case "y", "Y", "true":
+		ginkgoArgs = append(ginkgoArgs, "--p")
+		if len(skip) == 0 {
+			skip = serialTestsRegexp
+		}
+	}
+
+	ginkgoArgs = append(ginkgoArgs, []string{
+		"--focus=" + env.Getenv(focusEnvKey),
+		"--skip=" + skip,
+		"--noColor=true",
+	}...)
+
+	if len(env.Getenv(extraGinkgoArgsEnvKey)) > 0 {
+		ginkgoArgs = append(ginkgoArgs, env.Getenv(extraGinkgoArgsEnvKey))
+	}
+
+	extraArgs := []string{
+		"--disable-log-dump",
+		"--repo-root=/kubernetes",
+		"--provider=" + env.Getenv(providerEnvKey),
+		"--report-dir=" + env.Getenv(resultsDirEnvKey),
+		"--kubeconfig=" + env.Getenv(kubeconfigEnvKey),
+	}
+
+	// Extra args handling
+	sep := " "
+	if len(extraArgsSeparaterEnvKey) > 0 {
+		sep = env.Getenv(extraArgsSeparaterEnvKey)
+	}
+
+	if len(env.Getenv(extraGinkgoArgsEnvKey)) > 0 {
+		ginkgoArgs = append(ginkgoArgs, strings.Split(env.Getenv(extraGinkgoArgsEnvKey), sep)...)
+	}
+
+	if len(env.Getenv(extraArgsEnvKey)) > 0 {
+		extraArgs = append(extraArgs, strings.Split(env.Getenv(extraArgsEnvKey), sep)...)
+	}
+
+	if len(env.Getenv(dryRunEnvKey)) > 0 {
+		ginkgoArgs = append(ginkgoArgs, "--dryRun=true")
+	}
+
+	args := []string{}
+	args = append(args, ginkgoArgs...)
+	args = append(args, env.Getenv(testBinEnvKey))
+	args = append(args, "--")
+	args = append(args, extraArgs...)
+
+	cmd := exec.Command(env.Getenv(ginkgoEnvKey), args...)
```

Comment:
So if we're doing a native runner, is there an easier way then shelling?

Reply:
> an easier way then shelling

We have things at a bare bones level here: no dependence on any other tool and arent calling out to a shell. I don't know what simplification I could shoot for here.

Do you just mean that handling all the env vars seems verbose? If so, then I would agree that it seems verbose to read but that this section of the code is pretty much 1-1 with the bash script but we also get all the new benefits:
 - solves the raciness/flakiness of w.r.t. the pid of the process
 - generic flags, unblocking anyone who wants to try a unique test setup
 - os independence of this functionality

Reply:
I suppose this is fine for now... I'm just of the opinion that in the fullness of time we should simplify the UX of running the tests directly vs. wrap-running.

Reply:
/shrug

I guess I'm unsure what the goal/result would be. That would mean the conformance image would be calling ginkgo directly and the user would be responsible for setting all the necessary flags.

That would be fine but gets away from the norm that seems to be in place of using env vars to configure containers. In addition, you'd have a really messy time to run your command and report results to Sonobuoy (I've done it, but you have to make a really messy compound command). I feel like there'd even be another hurdle in there somewhere.

Reply:
/unshrug

Apparently, that is a command :) I didn't realize the bot would take a shrug action.